code/modules/research/nanites/nanite_programs.dm

```diff
@@ -0,0 +1,257 @@
+	var/timer = 0 					//Seconds before the timer effect activates. Starts counting AFTER the activation delay
+	var/timer_type = NANITE_TIMER_DEACTIVATE //What happens when the timer runs out
+
+	//Signal codes, these handle remote input to the nanites. If set to 0 they'll ignore signals.
+	var/activation_code 	= 0 	//Code that activates the program [1-9999]
+	var/deactivation_code 	= 0 	//Code that deactivates the program [1-9999]
+	var/kill_code 			= 0		//Code that permanently removes the program [1-9999]
+	var/trigger_code 		= 0 	//Code that triggers the program (if available) [1-9999]
+
+	//Extra settings
+	//Must be listed in text form, with the same title they'll be displayed in the programmer UI
+	//Changing these values is handled by set_extra_setting()
+	//Viewing these values is handled by get_extra_setting()
+	//Copying these values is handled by copy_extra_settings_to()
+	var/list/extra_settings = list()
+
+/datum/nanite_program/triggered
+	use_rate = 0
+	trigger_cost = 5
+	trigger_cooldown = 50
+	can_trigger = TRUE
+
+/datum/nanite_program/Destroy()
+	if(host_mob)
+		if(activated)
+			deactivate()
+		if(passive_enabled)
+			disable_passive_effect()
+	if(nanites)
+		nanites.programs -= src
+	. = ..()
+
+/datum/nanite_program/proc/copy()
+	var/datum/nanite_program/new_program = new type()
+
+	new_program.activated = activated
+	new_program.activation_delay = activation_delay
+	new_program.timer = timer
+	new_program.timer_type = timer_type
+	new_program.activation_code = activation_code
+	new_program.deactivation_code = deactivation_code
+	new_program.kill_code = kill_code
+	new_program.trigger_code = trigger_code
+	copy_extra_settings_to(new_program)
+
+	return new_program
+
+/datum/nanite_program/proc/copy_programming(datum/nanite_program/target, copy_activated = TRUE)
+	if(copy_activated)
+		target.activated = activated
+	target.activation_delay = activation_delay
+	target.timer = timer
+	target.timer_type = timer_type
+	target.activation_code = activation_code
+	target.deactivation_code = deactivation_code
+	target.kill_code = kill_code
+	target.trigger_code = trigger_code
+	copy_extra_settings_to(target)
+
+/datum/nanite_program/proc/set_extra_setting(user, setting)
+	return
+
+/datum/nanite_program/proc/get_extra_setting(setting)
+	return
+
+/datum/nanite_program/proc/copy_extra_settings_to(datum/nanite_program/target)
+	return
+
+/datum/nanite_program/proc/on_add(datum/component/nanites/_nanites)
+	nanites = _nanites
+	if(nanites.host_mob)
+		on_mob_add()
+
+/datum/nanite_program/proc/on_mob_add()
+	host_mob = nanites.host_mob
+	if(activated) //apply activation effects if it starts active
+		activate()
+
+/datum/nanite_program/proc/toggle()
+	if(!activated)
+		activate()
+	else
+		deactivate()
+
+/datum/nanite_program/proc/activate()
+	activated = TRUE
+
+/datum/nanite_program/proc/deactivate()
+	if(passive_enabled)
+		disable_passive_effect()
+	activated = FALSE
+
+/datum/nanite_program/proc/on_process()
+	if(!host_mob)
```

Comment:
Just stop processing when you're not on a mob

Reply:
i already do that, this is more of a sanity check in case of deletion or something

Reply:
dont, let it runtime